Add operator support for dynamic quant on mobile

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32479 Add operator support for dynamic quant on mobile**

Summary:
Run dynamic quantization on mobile (similar to FBGEMM). Currently only implemented on linear operator

Test Plan:
python test/test_quantized.py TestDynamicQuantizedLinear.test_qlinear

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D19542980](https://our.internmc.facebook.com/intern/diff/D19542980)

supriyar committed:
Add operator support for dynamic quant on mobile

Summary:
Run dynamic quantization on mobile (similar to FBGEMM). Currently only implemented on linear operator

Test Plan:
python test/test_quantized.py TestDynamicQuantizedLinear.test_qlinear

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 186ac5d9:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=186ac5d9714063253242acd2e2d931643aaaec66).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4313232) pytorch_linux_xenial_py3_clang5_asan_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=89229278) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=89229278))
<details>
<summary>
<code>Jan 22 03:38:28 RuntimeError: test_quantized failed!</code>
</summary>

```
Jan 22 03:38:28     #154 0x563db18f23bd in main /tmp/build/80754af9/python_1578429706181/work/Programs/python.c:69 
Jan 22 03:38:28     #155 0x7fd01df2982f in __libc_start_main /build/glibc-LK5gWL/glibc-2.23/csu/../csu/libc-start.c:291 
Jan 22 03:38:28     #156 0x563db19d1083 in _start /home/rdonnelly/mc/conda-bld/compilers_linux-64_1534865402226/work/.build/src/glibc-2.12.2/csu/../sysdeps/x86_64/elf/start.S:103 
Jan 22 03:38:28  
Jan 22 03:38:28 SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /var/lib/jenkins/workspace/caffe2/utils/threadpool/pthreadpool.cc:332:13 in  
Jan 22 03:38:28 Traceback (most recent call last): 
Jan 22 03:38:28   File "test/run_test.py", line 457, in <module> 
Jan 22 03:38:28     main() 
Jan 22 03:38:28   File "test/run_test.py", line 450, in main 
Jan 22 03:38:28     raise RuntimeError(message) 
Jan 22 03:38:28 RuntimeError: test_quantized failed! 
Jan 22 03:38:28 + cleanup 
Jan 22 03:38:28 + retcode=1 
Jan 22 03:38:28 + set +x 
Jan 22 03:38:28 =================== sccache compilation log =================== 
Jan 22 03:38:28 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 22 03:38:28 Compile requests                 0 
Jan 22 03:38:28 Compile requests executed        0 
Jan 22 03:38:28 Cache hits                       0 
Jan 22 03:38:28 Cache misses                     0 
Jan 22 03:38:28 Cache timeouts                   0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

supriyar committed:
Update on "Add operator support for dynamic quant on mobile"


Summary:
Run dynamic quantization on mobile (similar to FBGEMM). Currently only implemented on linear operator

Test Plan:
python test/test_quantized.py TestDynamicQuantizedLinear.test_qlinear

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Add operator support for dynamic quant on mobile"


Summary:
Run dynamic quantization on mobile (similar to FBGEMM). Currently only implemented on linear operator

Test Plan:
python test/test_quantized.py TestDynamicQuantizedLinear.test_qlinear

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

z-a-f commented on aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp:
```diff
@@ -204,17 +209,186 @@ class QLinearDynamicInt8 final : public torch::OperatorKernel {
+      float max,
+      int32_t qmin,
+      int32_t qmax) {
+    double scale =
```
```c++
TORCH_CHECK(min <= max);
TORCH_CHECK(qmin < qmax);
```

z-a-f commented on aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp:
```diff
@@ -204,17 +209,186 @@ class QLinearDynamicInt8 final : public torch::OperatorKernel {
+    double scale =
+        (std::max(max, 0.f) - std::min(min, 0.f)) / ((double)qmax - qmin);
+    if (scale == 0) {
+      scale = 0.1;
```
Is there a reason for this magic number?

supriyar replied:
FBGEMM uses this to avoid divide by 0 errors.

z-a-f replied:
The question is why 0.1? Why not 1.0 or `qmax-qmin`? I don't have preference one way or another, but this case happens only if in all zeros tensors, and I wonder if we should standardize it everywhere?

z-a-f commented on aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp:
```diff
@@ -204,17 +209,186 @@ class QLinearDynamicInt8 final : public torch::OperatorKernel {
+      scale = 0.1;
+    }
+    // If scale is 0, we arbitrary adjust the scale to 0.1
+    assert(scale > 0);
```
Use `TORCH_CHECK` instead of `assert`. Also, with the suggestions above, this is unnecessary.

z-a-f commented on aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp:
```diff
@@ -204,17 +209,186 @@ class QLinearDynamicInt8 final : public torch::OperatorKernel {
+    // We extend the [min, max] interval to ensure that it contains 0.
+    // Otherwise, we would not meet the requirement that 0 be an exactly
+    // representable value.
+    min = std::min(min, 0.f);
```
You are already having these checks in the `scale` computation. I suggest moving it up, and reusing.

z-a-f commented on aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp:
```diff
@@ -204,17 +209,186 @@ class QLinearDynamicInt8 final : public torch::OperatorKernel {
+    double initial_zero_point =
+        zero_point_from_min_error < zero_point_from_max_error
+        ? zero_point_from_min
+        : zero_point_from_max;
```
Did you mean to do

```c++
double zero_point_from_min_error = std::abs(qmin) - std::abs(min / scale);
double zero_point_from_max_error = std::abs(qmax) - std::abs(max / scale);
```

If not, this is not really an absolute error, and `initial_zero_point` will always be equal to `zero_point_from_max`

supriyar replied:
The arithmetic error on the zero point computed from min or max will depend on the value of individual terms. It will roughly be epsilon * (sum of absolute values of terms).
(This is inspired from what FBGEMM does today, as I didn't want the calculations to diverge)

raghuramank100 replied:
Here is the reference implementation used for static quant (doesn't need to align with this implementation though)
```
                scale = (max_val - min_val) / float(qmax - qmin)
                scale = max(scale, self.eps)
                zero_point = qmin - round(min_val / scale)
                zero_point = max(qmin, zero_point)
                zero_point = min(qmax, zero_point)
                zero_point = int(zero_point)
```

z-a-f replied:
> The arithmetic error on the zero point computed from min or max will depend on the value of individual terms. It will roughly be epsilon * (sum of absolute values of terms).
> (This is inspired from what FBGEMM does today, as I didn't want the calculations to diverge)

I understand that you don't want to diverge from FBGEMM, but these lines don't make sense -- because `qmax` is guaranteed to be greater than `qmin`, and `max` is guarantted to be greater or equal than `min`, `initial_zero_point` will always be `zero_point_from_max_error`. I am almost positive FBGEMM meant to do `std::abs(qmin) - std::abs(min / scale)` instead of `+`. It is better to send a fix suggestion to FBGEMM instead of propagating the bug here.

z-a-f replied:
cc @jianyuh to the discussion

z-a-f commented on aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp:
```diff
@@ -204,17 +209,186 @@ class QLinearDynamicInt8 final : public torch::OperatorKernel {
+      nudged_zero_point = qmax;
+    } else {
+      nudged_zero_point = nearbyint(initial_zero_point);
+    }
```
Can you give an example when this part is necessary? I cannot see the necessity of nudging the ZP if we picked the scale properly.

supriyar replied:
It is possible for the initial_zero_point computed to be non-integral. let's say min is 5.5 and max is 555.5 - In this case the computed zero point will have to be nudged to get an integral value.

z-a-f replied:
Not really, the nudging cases will only be necessary if we don't have guarantees of zero being representable, but we do (kinda). Here is my logic:

Suppose `initial_zero_point = qmin - min / scale`. We guarantee that `min` is either negative or zero, which makes `initial_zero_point < qmin` to be always false. If we suppose that `initial_zero_point = qmax - max / scale`, the logic is the same: `max` is either positive or zero. Hence, `initial_zero_point > qmax` is always false, so the second condition in the if statement will never be executed.

We can discuss it offline :)

z-a-f commented on aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp:
```diff
@@ -204,17 +209,186 @@ class QLinearDynamicInt8 final : public torch::OperatorKernel {
+      if (curr > x_max) {
+        x_max = curr;
+      }
+    }
```
Can we use member methods, like `.min()` and `.max()` to get the values?

z-a-f commented on aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp:
```diff
@@ -204,17 +209,186 @@ class QLinearDynamicInt8 final : public torch::OperatorKernel {
+        /*min=*/x_min,
+        /*max=*/x_max,
+        0,
+        255);
```
```
/*qmin=*/0,
/*qmax=*/255);
```

z-a-f reviewed (changes requested):
Should we also add benchmarks under `r/benchmarks/operator_benchmark/pt/`?

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp:
```diff
@@ -204,17 +209,186 @@ class QLinearDynamicInt8 final : public torch::OperatorKernel {
+    std::int32_t zero_point;
+  };
+
+  qparams chooseQuantizationParams(
```
This function needs to be identical to the FBGEMM implementation so that numerics are consistent between server and mobile. Is there a reason we cannot reuse the same code?

supriyar replied:
Yes, it is using most of the logic from there. However the code for fbgemm is not available on mobile so we have to repurpose it here for our use case.

raghuramank100 replied:
This is not a aten op, so wondering if it is possible to have a test that checks this implementation for consistency with fbgemm::choose_quantization_params

supriyar replied:
@raghuramank100 Thanks, I'll work on sharing the implementation.

supriyar commented:
> Should we also add benchmarks under `r/benchmarks/operator_benchmark/pt/`?

It wouldn't run them on mobile, so I don't see the point. Unless you meant to benchmark the mobile operators on server.

supriyar committed:
Update on "Add operator support for dynamic quant on mobile"


Summary:
Run dynamic quantization on mobile (similar to FBGEMM). Currently only implemented on linear operator

Test Plan:
python test/test_quantized.py TestDynamicQuantizedLinear.test_qlinear

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on test/test_quantized.py:
```diff
@@ -1055,109 +1055,121 @@ class TestDynamicQuantizedLinear(TestCase):
+                return
+            use_channelwise = False
+            use_multi_dim_input = False
+            use_relu = False
```
Can we add support for FcRelu to be consistent with server implementation?

supriyar replied:
I'll add it in a subsequent PR. We will need to modify the kernels as well to add this.

raghuramank100 commented on test/test_quantized.py:
```diff
@@ -1055,109 +1055,121 @@ class TestDynamicQuantizedLinear(TestCase):
-            b_fp32 = torch.from_numpy(
-                _dequantize(b_q0, X_scale * int(W_scales[0].item()), 0)
-            ).to(dtype=torch.float) if use_bias else None
+            avoid_vpmaddubsw_overflow_linear(
```
For qnnpack, this condition need not be checked as it is specific to FBGEMM

supriyar committed:
Update on "Add operator support for dynamic quant on mobile"


Summary:
Run dynamic quantization on mobile (similar to FBGEMM). Currently only implemented on linear operator

Test Plan:
python test/test_quantized.py TestDynamicQuantizedLinear.test_qlinear

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Add operator support for dynamic quant on mobile"


Summary:
Run dynamic quantization on mobile (similar to FBGEMM). Currently only implemented on linear operator

Test Plan:
python test/test_quantized.py TestDynamicQuantizedLinear.test_qlinear

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D19542980](https://our.internmc.facebook.com/intern/diff/D19542980)

[ghstack-poisoned]

jianyuh commented on aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp:
```diff
@@ -73,14 +76,13 @@ class QLinearDynamicInt8 final : public torch::OperatorKernel {
     // ReQuantizeForFloat won't index past 0.
 
-    const float* bias_ptr = nullptr;
+    float* bias_ptr = nullptr;
```
Any reason `const` is removed?

jianyuh commented on aten/src/ATen/native/quantized/cpu/quant_utils.h:
```diff
@@ -0,0 +1,117 @@
+  if (scale == 0.0f || isinf(1.0f / scale)) {
+    scale = 0.1;
+  }
+  TORCH_CHECK(scale > 0, "qlinear_dynamic scale should be > 0");
```
`ChooseQuantizationParams` is not only used by `qlinear_dynamic`. Maybe just specify "scale should be > 0" here.

jianyuh commented on test/test_quantized.py:
```diff
@@ -1055,109 +1055,120 @@ class TestDynamicQuantizedLinear(TestCase):
-            b_fp32 = torch.from_numpy(
-                _dequantize(b_q0, X_scale * int(W_scales[0].item()), 0)
-            ).to(dtype=torch.float) if use_bias else None
+            avoid_vpmaddubsw_overflow_linear(
```
Right: we shouldn't need to run this function (`avoid_vpmaddubsw_overflow_linear`) as there is no saturation issues for QNNPACK raised from `vpmaddubsw` instruction on x86.

jianyuh commented on test/test_quantized.py:
```diff
@@ -1055,109 +1055,120 @@ class TestDynamicQuantizedLinear(TestCase):
+
+            # Observe X_fp32 and determine X_scale and X_zero_point, this should match
+            # internals of dynamic linear.
+            X_scale, X_zp = _calculate_dynamic_qparams(X_fp32, torch.quint8)
```
If we enable the `reduce_range` option in the future, we also need to do the corresponding changes to `_calculate_dynamic_qparams` (https://github.com/pytorch/pytorch/blob/f050b16dd95b2bcce9853882fd3fb07a6fd80378/torch/testing/_internal/common_quantized.py#L41) in order to facilitate the unit tests

supriyar replied:
Good catch! I'll update this

jianyuh reviewed (commented):
Thanks for working on this!

jianyuh commented on aten/src/ATen/native/quantized/cpu/quant_utils.h:
```diff
@@ -0,0 +1,117 @@
+  int precision;
+};
+
+inline TensorQuantizationParams ChooseQuantizationParams(
```
Want to point out that ChooseQuantizationParams in FBGEMM (https://github.com/pytorch/FBGEMM/blob/master/src/QuantUtils.cc#L19) is originally from GEMMLowp:
https://github.com/google/gemmlowp/blob/master/doc/quantization_example.cc#L71

supriyar committed:
Update on "Add operator support for dynamic quant on mobile"


Summary:
Run dynamic quantization on mobile (similar to FBGEMM). Currently only implemented on linear operator

Test Plan:
python test/test_quantized.py TestDynamicQuantizedLinear.test_qlinear

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D19542980](https://our.internmc.facebook.com/intern/diff/D19542980)

[ghstack-poisoned]

jianyuh commented on aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp:
```diff
@@ -73,7 +76,6 @@ class QLinearDynamicInt8 final : public torch::OperatorKernel {
         /*preserve_sparsity=*/false);
 
     q_params.precision = precision;
-
```
nit: maybe keep this empty line?

jianyuh reviewed (approved):
Looks good to me!

supriyar committed:
Update on "Add operator support for dynamic quant on mobile"


Summary:
Run dynamic quantization on mobile (similar to FBGEMM). Currently only implemented on linear operator

Test Plan:
python test/test_quantized.py TestDynamicQuantizedLinear.test_qlinear

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D19542980](https://our.internmc.facebook.com/intern/diff/D19542980)

[ghstack-poisoned]

supriyar commented:
@pytorchbot retest this please

supriyar committed:
Update on "Add operator support for dynamic quant on mobile"


Summary:
Run dynamic quantization on mobile (similar to FBGEMM). Currently only implemented on linear operator

Test Plan:
python test/test_quantized.py TestDynamicQuantizedLinear.test_qlinear

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D19542980](https://our.internmc.facebook.com/intern/diff/D19542980)

[ghstack-poisoned]

supriyar committed:
Update on "Add operator support for dynamic quant on mobile"


Summary:
Run dynamic quantization on mobile (similar to FBGEMM). Currently only implemented on linear operator

Test Plan:
python test/test_quantized.py TestDynamicQuantizedLinear.test_qlinear

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D19542980](https://our.internmc.facebook.com/intern/diff/D19542980)

[ghstack-poisoned]

z-a-f commented on aten/src/ATen/native/quantized/cpu/quant_utils.h:
```diff
@@ -0,0 +1,118 @@
+    nudged_zero_point = qmax;
+  } else {
+    nudged_zero_point = nearbyint(initial_zero_point);
+  }
```
Are we still nudging? I thought I explained why this is not necessary
